File: kraken/src/main/java/glitch/kraken/services/ChannelService.java
package glitch.kraken.services;

import glitch.api.http.HttpRequest;
import glitch.api.http.Routes;
import glitch.api.objects.json.interfaces.OrdinalList;
import glitch.auth.GlitchScope;
import glitch.auth.objects.json.Credential;
import glitch.exceptions.http.ResponseException;
import glitch.kraken.GlitchKraken;
import glitch.kraken.object.json.*;
import glitch.kraken.object.json.collections.Communities;
import glitch.kraken.object.json.collections.Editors;
import glitch.kraken.object.json.collections.Teams;
import glitch.kraken.object.json.impl.AuthChannelImpl;
import glitch.kraken.services.request.ChannelSubscribersRequest;
import glitch.kraken.services.request.ChannelUserFollowsRequest;
import glitch.kraken.services.request.ChannelVideosRequest;
import glitch.service.AbstractHttpService;
import java.util.*;
import java.util.stream.Collectors;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

public class ChannelService extends AbstractHttpService {

    private static final Set<Integer> commercialDuration = new LinkedHashSet<>(Arrays.asList(30, 60, 90, 120, 150, 180));

    public ChannelService(GlitchKraken rest) {
        super(rest.getClient(), rest.getHttpClient());
    }

    public Mono<AuthChannel> getChannel(Credential credential) {
        return Mono.just(checkRequiredScope(credential.getScopes(), GlitchScope.CHANNEL_READ)).flatMap(b -> {
            if (b) {
                return exchangeTo(Routes.get("/channel").newRequest()
                                .header("Authorization", "OAuth " + credential.getAccessToken()),
                        AuthChannelImpl.class)
                        .cast(AuthChannel.class);
            } else {
                return Mono.error(handleScopeMissing(GlitchScope.CHANNEL_READ));
            }
        });
    }

    public Mono<Channel> getChannel(Long id) {
        return exchangeTo(Routes.get("/channels/%s").newRequest(id), Channel.class);
    }

    public Mono<Channel> updateChannel(Credential credential, ChannelBody data) {
        return updateChannel(credential.getUserId(), credential, data);
    }

    public Mono<Channel> updateChannel(Long id, Credential credential, ChannelBody body) {
        return Mono.just(checkRequiredScope(credential.getScopes(), GlitchScope.CHANNEL_EDITOR)).flatMap(b -> {
            if (b) {
                ChannelBody channelBody = (!id.equals(credential.getUserId())) ?
                        // preventing updates if credential is not channel owner
                        ChannelBody.builder(body)
                                .setDelay(null)
                                .setChannelFeedEnabled(null)
                                .build()
                        : body;

                return exchangeTo(Routes.put("/channels/%s").newRequest(id)
                        .header("Authorization", "OAuth " + credential.getAccessToken())
                        .body(HttpRequest.BodyType.JSON, Collections.singletonMap("channel", channelBody)), Channel.class);
            } else {
                return Mono.error(handleScopeMissing(GlitchScope.CHANNEL_EDITOR));
            }
        });
    }

    public Flux<User> getChannelEditors(Long id, Credential credential) {
        return Mono.just(checkRequiredScope(credential.getScopes(), GlitchScope.CHANNEL_READ))
                .flatMapMany(b -> {
                    if (b) {
                        return exchangeTo(Routes.get("/channels/%s").newRequest(id)
                                .header("Authorization", "OAuth " + credential.getAccessToken()), Editors.class)
                                .flatMapIterable(OrdinalList::getData);
                    } else {
                        return Mono.error(handleScopeMissing(GlitchScope.CHANNEL_READ));
                    }
                });
    }

    public ChannelUserFollowsRequest getChannelFollowers(Long id) {
        return new ChannelUserFollowsRequest(http, id);
    }

    public Flux<Team> getChannelTeams(Long id) {
        return exchangeTo(Routes.get("/channels/%s/teams").newRequest(id), Teams.class)
                .flatMapIterable(OrdinalList::getData);
    }

    public ChannelSubscribersRequest getChannelSubscribers(Long id, Credential credential) {
        return new ChannelSubscribersRequest(http, id, credential);
    }

    public Mono<Subscriber> checkChannelSubscriptionByUser(Channel channel, User user, Credential credential) {
        return checkChannelSubscriptionByUser(channel.getId(), user.getId(), credential);
    }

    public Mono<Subscriber> checkChannelSubscriptionByUser(Long channelId, Long userId, Credential credential) {

        HttpRequest request = Routes.get("/channels/%s/subscriptions/%s").newRequest(channelId, userId);

        return Mono.just(checkRequiredScope(credential.getScopes(), GlitchScope.CHANNEL_CHECK_SUBSCRIPTION))
                .flatMap(b -> {
                    if (b) {
                        return exchangeTo(request, Subscriber.class).onErrorResume(ResponseException.class, (e) -> {
                            if (e.getStatus() == 404 && e.getStatusMessage().matches("^(.+) has no subscriptions to (.+)$")) {
                                return Mono.error(new NullPointerException(e.getStatusMessage()));
                            } else {
                                return Mono.error(e);
                            }
                        });
                    } else {
                        return Mono.error(handleScopeMissing(GlitchScope.CHANNEL_CHECK_SUBSCRIPTION));
                    }
                });
    }

    public ChannelVideosRequest getChannelVideos(Long id) {
        return new ChannelVideosRequest(http, id);
    }

    public Mono<CommercialData> startChannelCommercial(Long id, Credential credential, int duration) {
        Map<String, Object> dur = Collections.singletonMap("length", commercialDuration.stream().min(Comparator.comparingInt(i -> Math.abs(i - duration))).orElse(30));

        return Mono.just(checkRequiredScope(credential.getScopes(), GlitchScope.CHANNEL_COMMERCIAL))
                .flatMap(b -> {
                    if (b) {
                        return exchangeTo(Routes.post("/channels/%s/commercial").newRequest(id).body(HttpRequest.BodyType.JSON, dur)
                                .header("Authorization", "OAuth " + credential.getAccessToken()), CommercialData.class);
                    } else {
                        return Mono.error(handleScopeMissing(GlitchScope.CHANNEL_COMMERCIAL));
                    }
                });
    }

    public Mono<AuthChannel> resetStreamKey(Long id, Credential credential) {
        return Mono.just(checkRequiredScope(credential.getScopes(), GlitchScope.CHANNEL_STREAM))
                .flatMap(b -> {
                    if (b) {
                        return exchangeTo(Routes.delete("/channels/%s/stream_key").newRequest(id)
                                .header("Authorization", "OAuth " + credential.getAccessToken()), AuthChannel.class);
                    } else {
                        return Mono.error(handleScopeMissing(GlitchScope.CHANNEL_STREAM));
                    }
                });
    }

    public Flux<Community> getChannelCommunities(Long id) {
        return exchangeTo(Routes.get("/channels/%s/communities").newRequest(id), Communities.class)
                .flatMapIterable(OrdinalList::getData);
    }

    public Mono<Boolean> setChannelCommunities(Long id, Credential credential, Collection<UUID> communityIds) {
        return Mono.just(checkRequiredScope(credential.getScopes(), GlitchScope.CHANNEL_EDITOR))
                .flatMap(b -> {
                    if (b) {
                        return exchange(Routes.put("/channels/%s/communities").newRequest(id)
                                .body(HttpRequest.BodyType.JSON, Collections.singletonMap("community_ids", communityIds.stream().map(UUID::toString).collect(Collectors.toList())))
                                .header("Authorization", "OAuth " + credential.getAccessToken()))
                                .map(response -> response.getStatus().getCode() == 204);
                    } else {
                        return Mono.error(handleScopeMissing(GlitchScope.CHANNEL_EDITOR));
                    }
                });
    }

    public Mono<Boolean> clearChannelCommunities(Long id, Credential credential) {
        return Mono.just(checkRequiredScope(credential.getScopes(), GlitchScope.CHANNEL_EDITOR))
                .flatMap(b -> {
                    if (b) {
                        return exchange(Routes.delete("/channels/%s/community").newRequest(id)
                                .header("Authorization", "OAuth " + credential.getAccessToken()))
                                .map(response -> response.getStatus().getCode() == 204);
                    } else {
                        return Mono.error(handleScopeMissing(GlitchScope.CHANNEL_EDITOR));
                    }
                });
    }
}

File: kraken/src/main/kotlin/io/glitchlib/v5/service/ChannelService.kt
package io.glitchlib.v5.service

import io.glitchlib.GlitchClient
import io.glitchlib.auth.Credential
import io.glitchlib.auth.Scope
import io.glitchlib.internal.http.bodyFlowable
import io.glitchlib.internal.http.bodySingle
import io.glitchlib.internal.http.completed
import io.glitchlib.model.CursorList
import io.glitchlib.model.OrdinalList
import io.glitchlib.v5.internal.AbstractKrakenService
import io.glitchlib.v5.internal.model.json.AuthChannelImpl
import io.glitchlib.v5.internal.model.json.ChannelFollows
import io.glitchlib.v5.internal.model.json.ChannelImpl
import io.glitchlib.v5.internal.model.json.Editors
import io.glitchlib.v5.model.json.AuthChannel
import io.glitchlib.v5.model.json.Channel
import io.glitchlib.v5.model.json.ChannelBody
import io.glitchlib.v5.model.json.CommercialData
import io.glitchlib.v5.model.json.Community
import io.glitchlib.v5.model.json.Subscriber
import io.glitchlib.v5.model.json.TeamItem
import io.glitchlib.v5.model.json.Video
import io.glitchlib.v5.service.request.ChannelFollowRequest
import io.glitchlib.v5.service.request.ChannelSubscribersRequest
import io.glitchlib.v5.service.request.ChannelVideoRequest
import io.glitchlib.v5.service.request.CollectionRequest
import io.reactivex.Maybe
import io.reactivex.Single
import io.reactivex.rxkotlin.cast
import java.util.*

class ChannelService internal constructor(client: GlitchClient) : AbstractKrakenService(client) {
    fun getChannel(credential: Credential) = if (credential.scopeCheck(Scope.CHANNEL_READ))
        get<AuthChannelImpl>("/channel") {
            addHeaders("Authorization", "OAuth ${credential.accessToken}")
        }.bodySingle.cast<AuthChannel>()
    else scopeIsMissing(Scope.CHANNEL_READ)

    fun getChannel(id: Long) =
            get<ChannelImpl>("/channels/$id").bodySingle.cast<Channel>()

    fun updateChannel(credential: Credential, id: Long, data: ChannelBody.() -> Unit = {}) =
            if (credential.scopeCheck(Scope.CHANNEL_EDITOR))
                ChannelBody().apply(data).apply {
                    if (id != credential.id) {
                        delay = null
                    }
                }.let {
                    put<AuthChannelImpl>("/channels/$id") {
                        addHeaders("Authorization", "OAuth ${credential.accessToken}")
                        setBody(mapOf("channel" to it))
                    }.bodySingle.cast<AuthChannel>()
                }
            else scopeIsMissing(Scope.CHANNEL_EDITOR)

    fun getChannelEditors(credential: Credential, id: Long) =
            if (credential.scopeCheck(Scope.CHANNEL_READ))
                get<Editors>("/channels/$id/editors") {
                    addHeaders("Authorization", "OAuth ${credential.accessToken}")
                }.bodySingle
            else scopeIsMissing(Scope.CHANNEL_READ)

    fun getChannelFollows(id: Long, request: ChannelFollowRequest.() -> Unit = {}) =
            get<ChannelFollows>("/channels/$id/editors", ChannelFollowRequest().apply(request)()).bodySingle

    fun getChannelTeams(id: Long) =
            get<OrdinalList<TeamItem>>("/channels/$id/teams").bodyFlowable

    fun getChannelSubscribers(credential: Credential, id: Long, request: ChannelSubscribersRequest.() -> Unit = {}) =
            if (credential.scopeCheck(Scope.CHANNEL_SUBSCRIPTIONS))
                get<OrdinalList<Subscriber>>(
                        "/channels/$id/subscriptions",
                        ChannelSubscribersRequest(credential).apply(request)()
                ).bodyFlowable
            else scopeIsMissing<Subscriber>(Scope.CHANNEL_SUBSCRIPTIONS).toFlowable()

    fun checkChannelSubscriptionByUser(credential: Credential, channelId: Long, userId: Long) =
            if (credential.scopeCheck(Scope.CHANNEL_SUBSCRIPTIONS))
                get<Subscriber>("/channels/$channelId/subscriptions/$userId") {
                    addHeaders("Authorization", "OAuth ${credential.accessToken}")
                }.flatMapMaybe {
                    if (it.status.isError && it.status.code == 404) Maybe.error(NullPointerException(it.status.message))
                    else it.body
                } else scopeIsMissing<Subscriber>(Scope.CHANNEL_SUBSCRIPTIONS).toMaybe()

    fun getChannelVideos(id: Long, request: ChannelVideoRequest.() -> Unit = {}) =
            get<OrdinalList<Video>>("/channels/$id/videos", ChannelVideoRequest().apply(request)())

    fun startChannelCommercial(id: Long, credential: Credential, duration: Int) =
            if (credential.scopeCheck(Scope.CHANNEL_COMMERCIAL))
                post<CommercialData>("/channels/$id/commercial") {
                    setBody(mapOf("length" to (COMMERCIAL_DURATION.minBy { Math.abs(it - duration) } ?: 30).toString()))
                    addHeaders("Authorization", "OAuth ${credential.accessToken}")
                }.bodySingle
            else scopeIsMissing<CommercialData>(Scope.CHANNEL_COMMERCIAL)

    fun resetStreamKey(id: Long, credential: Credential): Single<AuthChannel> =
            if (credential.scopeCheck(Scope.CHANNEL_STREAM))
                delete<AuthChannelImpl>("/channels/$id/stream_key") {
                    addHeaders("Authorization", "OAuth ${credential.accessToken}")
                }.bodySingle.cast()
            else scopeIsMissing(Scope.CHANNEL_STREAM)

    fun getChannelCommunities(id: Long) =
            get<OrdinalList<Community>>("/channels/$id/communities").bodyFlowable

    fun setChannelCommunities(id: Long, credential: Credential, vararg communityIds: String) =
            setChannelCommunities(id, credential, *communityIds.map(UUID::fromString).toTypedArray())

    fun setChannelCommunities(id: Long, credential: Credential, vararg communityIds: UUID) =
            setChannelCommunities(id, credential, communityIds.toSet())

    fun setChannelCommunities(id: Long, credential: Credential, communityIds: Collection<UUID>) =
            if (credential.scopeCheck(Scope.CHANNEL_EDITOR))
                put<Unit>("/channels/$id/communities") {
                    setBody(mapOf("community_ids" to communityIds.map { it.toString() }.toSet()))
                    addHeaders("Authorization", "OAuth ${credential.accessToken}")
                }.completed
            else scopeIsMissing<Unit>(Scope.CHANNEL_EDITOR).ignoreElement()

    fun clearChannelCommunities(id: Long, credential: Credential, communityIds: Collection<UUID>) =
            if (credential.scopeCheck(Scope.CHANNEL_EDITOR))
                delete<Unit>("/channels/$id/communities") {
                    addHeaders("Authorization", "OAuth ${credential.accessToken}")
                }.completed
            else scopeIsMissing<Unit>(Scope.CHANNEL_EDITOR).ignoreElement()

    fun getCollections(id: Long, request: CollectionRequest.() -> Unit = {}) =
            get<CursorList<io.glitchlib.v5.model.json.Collection>>(
                    "/channels/$id/collections",
                    CollectionRequest().apply(request).invoke()
            ).bodySingle

    private val COMMERCIAL_DURATION = setOf(30, 60, 90, 120, 150, 180)
}
